wrappers/python/nntile/layer/batch_norm.py

```diff
@@ -0,0 +1,253 @@
+        y = TensorMoments(
+            zeros(x.value.shape, dtype=type(x.value)),
+            zeros(x.value.shape, dtype=type(x.value)),
+            grad_required=True,
+        )
+        n_channels = x.value.shape[1]
+
+        weight = TensorMoments(
+            ones([n_channels], dtype=type(x.value)),
+            ones([n_channels], dtype=type(x.value)),
+            grad_required=True,
+        )
+        bias = TensorMoments(
+            ones([n_channels], dtype=type(x.value)),
+            ones([n_channels], dtype=type(x.value)),
+            grad_required=True,
+        )
+
+        return cls(x, y, weight, bias, eps=eps, redux=redux)
+
+    def forward_async(self):
+        self._normalize_forward()
+        self._learnable_transform_forward()
+
+        self.y.value.wont_use()
+
+    def backward_async(self):
+        self._learnable_transform_backward()
+        self._normalize_backward()
+
+        # could be earlier. But this is straightforward we delete grad
+        self.x.grad.wont_use()
+        self.y.grad.wont_use()
+
+    def _normalize_forward(self):
+        # TODO: add add_fiber3_async and remove copy
+        copy_async(self.x.value, self.x_normalized)
+
+        # mean
+        x_mean = self.tmp_buff_channels
+        sum_fiber_async(1.0 / self.numel_in_channel, self.x.value, 0.0, x_mean, 1, 0)
+        self.x.value.wont_use()
+
+        # X_res = X - mean
+        add_fiber_async(-1.0, x_mean, 1.0, self.x_normalized, 1, 0)
+        # copy for backward
+        copy_async(self.x_normalized, self.x_unbiased_copy)
+
+        # inverse std
+        norm_slice_async(
+            1.0, self.x_normalized, 0.0, self.std_tmp_3dim, 0, redux=self.redux
+        )
+        norm_slice_async(
+            1.0, self.std_tmp_3dim, 0.0, self.std_tmp_2dim, 2, redux=self.redux
+        )
+        self.std_tmp_3dim.invalidate_submit()
+        norm_slice_async(
+            1.0 / self.numel_in_channel**0.5,
+            self.std_tmp_2dim,
+            0.0,
+            self.inv_std,
+            1,
+            redux=self.redux,
+        )
+        self.std_tmp_2dim.invalidate_submit()
+
+        hypot_scalar_inverse_async(self.eps, 1.0, self.inv_std)
+
+        # X_res = X_res/std
+        prod_fiber_async(self.inv_std, 1.0, self.x_normalized, 1)
+
+    def _learnable_transform_forward(self):
+        # y = weight * y + bias
+        copy_async(self.x_normalized, self.y.value)
+        prod_fiber_async(self.weight.value, 1.0, self.y.value, 1)
+        add_fiber_async(1.0, self.bias.value, 1.0, self.y.value, 1, 0)
+        self.bias.value.wont_use()
+
+    def _compute_grad_normalized_input_over_x(self, A_grad_nnt):
+        grad_mean = self.tmp_buff_channels  # zeros([self.n_channels])
+        sum_fiber_async(1.0 / self.numel_in_channel, A_grad_nnt, 0.0, grad_mean, 1, 0)
+
+        grad_x_nnt = self.x.grad  # zeros(A_grad_nnt.shape)
+        copy_async(A_grad_nnt, grad_x_nnt)
```

Comment:
A_grad_nnt is not used after this point -> invalidate submit

Reply:
done